lib/managedservicesplatform/iam/database.go

```diff
@@ -0,0 +1,121 @@
+	"go.uber.org/zap"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+	gormlogger "gorm.io/gorm/logger"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/lib/redislock"
+)
+
+const databaseName = "msp-iam"
+
+type metadata struct {
+	AuthorizationModelID  string    `gorm:"column:authorization_model_id"`
+	AuthorizationModelDSL string    `gorm:"column:authorization_model_dsl"`
+	UpdatedAt             time.Time `gorm:"not null"`
+}
+
+// migrateAndRecon migrates the "msp-iam" database schema (when needed) and
+// recons the framework metadata.
+func migrateAndRecon(ctx context.Context, logger log.Logger, sqlDB *sql.DB, redisClient *redis.Client) (_ *metadata, err error) {
+	ctx, span := iamTracer.Start(ctx, "iam.migrateAndRecon")
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+		}
+		span.End()
+	}()
+
+	conn, err := gorm.Open(
+		postgres.New(postgres.Config{Conn: sqlDB}),
+		&gorm.Config{
+			SkipDefaultTransaction: true,
+			NowFunc: func() time.Time {
+				return time.Now().UTC().Truncate(time.Microsecond)
+			},
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "open connection")
+	}
+
+	goose.SetBaseFS(assets.EmbedMigrations)
+	goose.SetLogger(&gooseLoggerShim{Logger: logger})
+
+	currentVersion, err := goose.GetDBVersion(sqlDB)
+	if err != nil {
+		return nil, errors.Wrap(err, "get DB version")
+	}
+	span.SetAttributes(attribute.Int64("currentVersion", currentVersion))
+
+	// We want to make sure only one instance of the server is doing auto-migration
+	// at a time.
+	err = redislock.OnlyOne(
+		logger,
+		redisClient,
+		fmt.Sprintf("%s:auto-migrate", databaseName),
+		15*time.Second,
+		func() error {
+			span.AddEvent("lock.acquired")
+
+			// Create a session that ignore debug logging.
+			sess := conn.Session(&gorm.Session{
+				Logger: gormlogger.Default.LogMode(gormlogger.Warn),
+			})
+			// Auto-migrate database table definitions.
+			for _, table := range []any{&metadata{}} {
+				err := sess.AutoMigrate(table)
+				if err != nil {
+					return errors.Wrapf(err, "auto migrating table for %s", errors.Safe(fmt.Sprintf("%T", table)))
+				}
+			}
+
+			// Migrate OpenFGA's database schema.
+			err = goose.UpContext(
+				ctx,
+				sqlDB,
+				assets.PostgresMigrationDir,
+			)
+			if err != nil {
+				return errors.Wrap(err, "run migrations")
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "auto-migrate")
+	}
+
+	var md metadata
+	if err = conn.FirstOrCreate(&md).Error; err != nil {
+		return nil, errors.Wrap(err, "init metadata")
+	}
+	return &md, nil
+}
+
+type gooseLoggerShim struct {
+	log.Logger
+}
+
+func (l *gooseLoggerShim) Fatalf(format string, v ...interface{}) {
+	l.Fatal("goose", zap.String("message", fmt.Sprintf(format, v...)))
+}
+
+func (l *gooseLoggerShim) Printf(format string, v ...interface{}) {
+	l.Debug("goose", zap.String("message", fmt.Sprintf(format, v...)))
+}
```

Comment:
nit: `"goose"` should be scope, let's just put the message in the message already formatted

```suggestion
type gooseLoggerShim struct {
	log.Logger
}

func (l *gooseLoggerShim) Fatalf(format string, v ...interface{}) {
	l.Fatal(fmt.Sprintf(format, v...))
}

func (l *gooseLoggerShim) Printf(format string, v ...interface{}) {
	l.Debug(fmt.Sprintf(format, v...))
}
```